packages/langium/src/utils/promise-util.ts

```diff
@@ -84,22 +84,25 @@ export async function interruptAndCheck(token: CancellationToken): Promise<void>
 export class MutexLock {
 
     private previousAction = Promise.resolve();
+    private previousTokenSource = new CancellationTokenSource();
 
     /**
      * Performs a single async action, like initializing the workspace or processing document changes.
      * Only one action will be executed at a time.
      */
-    async lock(action: () => Promise<void>): Promise<void> {
+    lock(action: (token: CancellationToken) => Promise<void>): Promise<void> {
+        this.previousTokenSource.cancel();
```

Comment:
You are cancelling the current task in progress? Or how to understand this?

Reply:
Yep, that the idea. We want to stop processing the current workspace change in favor of processing the newest changes instead.

Reply:
I see a bit a contradiction. Because you can use a Mutex everywhere. Does this mean if two asynchronous processes acquire the lock, the last one will always cancel the first one? I do not see any relation to a workspace HERE in this file. Because HERE is very generic and not WORKSPACE only.

Reply:
I don't really see a contradiction here:

1. We don't force to cancel the execution (by some v8 nodejs magic). We simply notify that a new action has arrived, and that the action can return from execution. They don't have to listen to this cancellation if they don't want to.
2. You can create mutexes for any use case you want, we use a single mutex for workspace writing activities. Another use case would use a separate mutex, which has it's own execution context.